### PR TITLE
fix(orm): strip NULL bytes from JSONB writes (#370)

### LINKED ISSUE #370: runs.output JSONB write fails with UntranslatableCharacterError when output contains a NULL byte
## Summary

When a LangGraph node returns state containing a literal NULL byte (`\u0000`, the U+0000 code point) in any string — for example, agent-synthesised Unicode test data or untrusted user-supplied input that happens to include `\x00` — `aegra_api.services.run_status.finalize_run` (and `update_run_status`) raises an `asyncpg.exceptions.UntranslatableCharacterError` on the `UPDATE runs SET output=$2::JSONB ...` statement. The run is marked `error` even though the graph itself completed successfully.

## Observed

```
(sqlalchemy.dialects.postgresql.asyncpg.Error) <class 'asyncpg.exceptions.UntranslatableCharacterError'>: unsupported Unicode escape sequence
DETAIL:  \u0000 cannot be converted to text.
[SQL: UPDATE runs SET status=$1::VARCHAR, output=$2::JSONB, updated_at=$3::TIMESTAMP WITH TIME ZONE WHERE runs.run_id = $4::VARCHAR]
```

The thread state survives because the LangGraph checkpointer writes to BYTEA columns via `JsonPlusSerializer` (msgpack-based, null-byte-safe). Only the JSONB `output` write breaks. The full agent output is still recoverable from `state.values.files[...]` even after the run is marked `error`.

## Root cause (against `aegra-api 0.9.4`)

1. **`aegra_api/core/serializers/general.py:49-50`** — `GeneralSerializer._serialize_object` returns `str` unchanged. No null-byte strip, no `ensure_ascii`, no control-character filter.
2. **`aegra_api/services/run_status.py:51` (`update_run_status`) and `:102` (`finalize_run`)** — both call `_safe_serialize(output, ...)` then execute the JSONB UPDATE.
3. **`_safe_serialize`'s try/except wraps only the serializer call**, not `session.execute()` / `commit()`. The asyncpg failure is raised outside the try block, so the run ends up `error` rather than degrading to the fallback dict.

## Scope

`aegra_api/core/orm.py` defines roughly 10 JSONB columns across the runs / threads / assistants tables (`output`, `input`, `context`, `metadata`). All of them inherit the same vulnerability — anything serialised into one is at risk if it contains `\u0000`.

## Suggested fix

Two options, in order of preference:

1. **`TypeDecorator` over `JSONB`** in `aegra_api/core/orm.py`. A small `JsonbNoNull` type that recursively walks the value and does `s.replace("\u0000", "")` (or replaces with `U+FFFD`). Covers all 10 columns uniformly; protects future JSONB additions automatically.

2. **Recursive `_strip_null_bytes(value)` helper inside `_safe_serialize`** (`run_status.py:113`). Smaller diff, narrower coverage.

In either case, **widen `_safe_serialize`'s try/except** to wrap the actual `session.execute()` so a future serialisation failure degrades to the fallback dict instead of marking the run `error`.

## Minimal repro

Any graph that produces state containing `"\u0000"` in a string will trigger this on the next run-status update. The simplest:

```python
from langgraph.graph import StateGraph, START, END
from typing import TypedDict

class S(TypedDict, total=False):
    out: str

def node(state):
    return {"out": "before\u0000after"}

g = StateGraph(S)
g.add_node("n", node)
g.add_edge(START, "n")
g.add_edge("n", END)
compiled = g.compile()
```

Deploy as an Aegra graph, invoke once via `POST /threads/{tid}/runs`. The graph completes; the runs UPDATE then fails with `UntranslatableCharacterError`. Status: `error`. The thread state (`/threads/{tid}/state`) still contains the full output.

## Upstream precedent

[`langchain-ai/langgraph#1878`](https://github.com/langchain-ai/langgraph/issues/1878) reported the same `UntranslatableCharacterError` against `AsyncPostgresSaver`; fixed upstream in `langgraph-checkpoint-postgres==1.0.9`. That fix covers the checkpointer's BYTEA write path. Aegra's JSONB write path is separate and unaffected.

## Notes

- `gh search issues --repo aegra/aegra` (open + closed) returned no prior reports for this symptom.
- Happy to send a PR if a maintainer can confirm the preferred approach (TypeDecorator vs in-serializer strip).

### PR DESCRIPTION
**Pull request by ibbybuilds**
## Summary

Fixes #370. Postgres JSONB rejects U+0000; asyncpg surfaces it as `UntranslatableCharacterError` and the run gets marked `error` even though the graph completed successfully.

Root cause: `GeneralSerializer` passes strings through untouched, so any agent state containing a literal NULL byte (untrusted input, model hallucination, agent-generated test data) crashes the `UPDATE runs SET output=$2::JSONB` statement. Every JSONB column on the ORM (~10 across `assistant`, `assistant_versions`, `thread`, `runs`) has the same exposure.

## Fix

Add a `JsonbSafe` TypeDecorator in `core/orm.py` that recursively strips `\x00` from strings and dict keys in `process_bind_param`, then swap `JSONB` -> `JsonbSafe` on every JSONB column declaration. The read path is untouched, so existing rows are unaffected.

Single defense at the DB boundary covers all current and future JSONB columns uniformly — no scattering strip helpers across services.

## Why this shape

- one place to look — the column type says what it does
- new JSONB columns are auto-protected
- `"\x00" in v` short-circuits before allocating on clean strings (the 99.99% case)
- non-string scalars and other control chars (`\n`, `\t`, `\x01`) pass through untouched — JSONB only rejects U+0000

## Tests

New `tests/unit/test_core/test_orm_jsonb_safe.py` — 17 cases covering:
- clean strings returned identical (identity check)
- NULL byte stripping in strings, dict values, dict keys, lists, tuples, deeply nested
- other control chars preserved
- non-string scalars passthrough (None, int, float, bool)
- `process_bind_param` strips on the TypeDecorator instance
- `cache_ok = True` (required for SQLAlchemy statement caching)

All 1027 existing unit tests still pass.

## Version

Bumps both packages 0.9.14 -> 0.9.15 (patch — bug fix, no breaking changes).

## Test plan

- [x] unit tests pass (1027 + 17 new)
- [x] ruff clean
- [ ] e2e: graph node returning `{"out": "a\x00b"}` completes `success`, output queryable, NULL stripped
- [ ] regression: cover `input` (request body), `metadata_json` (thread metadata), `context` (assistant context)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Prevented Postgres errors by sanitizing null (U+0000) characters in JSON fields used by assistants, versions, threads, and runs.

* **Tests**
  * Added comprehensive unit tests for null-byte handling, deep nesting, and key-collision scenarios in JSON payloads.

* **Chores**
  * Version bumped to 0.9.16.

[Review Change Stack](https://app.coderabbit.ai/change-stack/aegra/aegra/pull/376)

<h3>Greptile Summary</h3>

This PR adds a `JsonbSafe` SQLAlchemy `TypeDecorator` that recursively strips `\x00` (NULL bytes) from string values and dict keys before writing to any JSONB column, fixing a crash where Postgres raised `UntranslatableCharacterError` on agent-generated payloads containing literal NULL bytes.

- Introduces `_strip_null_bytes` with a depth guard at 200 levels (preventing `RecursionError` on adversarial nesting) and a key-collision warning log when distinct raw keys collapse to the same stripped key.
- Replaces every `JSONB` column declaration across `Assistant`, `AssistantVersion`, `Thread`, and `Run` with `JsonbSafe`.
- Adds 17 unit tests covering clean strings, deep nesting, scalars, key collisions, and the depth guard.

<h3>Confidence Score: 5/5</h3>

Safe to merge — the fix is well-scoped and both concerns from the prior review round are properly addressed.

All JSONB columns are uniformly protected at the ORM boundary. The depth guard at 200 sits comfortably below Python's frame limit while covering every realistic payload. Key collisions now emit a structured log warning rather than silently dropping data. Tests cover the full decision tree including edge cases, and existing tests are unaffected.

No files require special attention.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| libs/aegra-api/src/aegra_api/core/orm.py | Adds JsonbSafe TypeDecorator with depth-guarded, collision-warning recursive NULL-byte stripping; swaps JSONB → JsonbSafe across all 10 JSONB columns cleanly. |
| libs/aegra-api/tests/unit/test_core/test_orm_jsonb_safe.py | 17 unit tests covering string stripping, deep nesting, scalars, tuples-as-lists, key collision warning, and depth-guard behaviour — good coverage of the new logic. |

</details>

<sub>Reviews (4): Last reviewed commit: ["fix(orm): guard recursion depth and log ..."](https://github.com/aegra/aegra/commit/4aaac99feb601a060d942a741b34d9d9ae99da83) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=31755427)</sub>

### CODE REVIEW

File: libs/aegra-cli/pyproject.toml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [project]
 name = "aegra-cli"
-version = "0.9.14"
+version = "0.9.15"
```
[the remark this answers is not in the document]
Comment (ibbybuilds) [reply libs/aegra-cli/pyproject.toml]: skipping. per CLAUDE.md, `~=0.9.0` is the intentional pre-1.0 compatible-release constraint: it means `>=0.9.0, <0.10.0` so patch bumps don't require touching the constraint. only a minor bump (0.9.x → 0.10.0) would. this is a patch bump (0.9.14 → 0.9.15), so the constraint stays put.
